File: src/main/java/com/kurento/demo/cpbrazil/CpbWebRtc.java
/*
 * (C) Copyright 2014 Kurento (http://kurento.org/)
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Lesser General Public License
 * (LGPL) version 2.1 which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/lgpl-2.1.html
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 */
package com.kurento.demo.cpbrazil;

import static com.google.common.collect.Lists.newArrayList;

import java.net.URISyntaxException;
import java.util.Calendar;

import org.springframework.beans.factory.annotation.Autowired;

import com.kurento.demo.cpbrazil.CpbWindows.Windows;
import com.kurento.demo.cpbrazil.youtube.Videos;
import com.kurento.kmf.content.ContentCommand;
import com.kurento.kmf.content.ContentCommandResult;
import com.kurento.kmf.content.WebRtcContentHandler;
import com.kurento.kmf.content.WebRtcContentService;
import com.kurento.kmf.content.WebRtcContentSession;
import com.kurento.kmf.media.ChromaFilter;
import com.kurento.kmf.media.FaceOverlayFilter;
import com.kurento.kmf.media.GStreamerFilter;
import com.kurento.kmf.media.MediaApiConfiguration;
import com.kurento.kmf.media.MediaPipeline;
import com.kurento.kmf.media.PointerDetectorAdvFilter;
import com.kurento.kmf.media.PointerDetectorWindowMediaParam;
import com.kurento.kmf.media.RecorderEndpoint;
import com.kurento.kmf.media.WebRtcEndpoint;
import com.kurento.kmf.media.WindowParam;
import com.kurento.kmf.media.events.MediaEventListener;
import com.kurento.kmf.media.events.WindowInEvent;
import com.kurento.kmf.repository.RepositoryHttpRecorder;
import com.kurento.kmf.repository.RepositoryItem;

/**
 * Campus Party Brazil 2014 Kurento demo. This demo has the following pipeline:
 * 
 * WebRTC -> RateLimiter -> MirrorFilter -> PointerDetectorFilter ->
 * ChromaFilter -> FaceOverlayFilter -> Recorder
 * 
 * @author Boni García (bgarcia@gsyc.es)
 * @since 1.0.1
 */
@WebRtcContentService(path = "/cpbWebRtc/*")
public class CpbWebRtc extends WebRtcContentHandler {

	// MediaPipeline and MediaElements
	public MediaPipeline mediaPipeline;
	public GStreamerFilter mirrorFilter;
	public GStreamerFilter rateLimiter;
	public PointerDetectorAdvFilter pointerDetectorAdvFilter;
	public FaceOverlayFilter faceOverlayFilter;
	public ChromaFilter chromaFilter;
	public RecorderEndpoint recorderEndpoint;

	// Token to upload videos to Kurento Brazil Demo playlist
	private static final String PLAYLIST_TOKEN = "PL58tWS2XjtialwG-eWDYoFwQpHTd5vDEE";

	// Global demo elements
	public CpbWindows cpbWindows;
	public String itemId;
	public Windows activeWindow;
	private String handlerUrl;
	private String recorderUrl;
	private int mario = 1;
	private int count = 1;

	@Autowired
	private MediaApiConfiguration config;

	@Override
	public void onContentRequest(final WebRtcContentSession contentSession)
			throws Exception {
		String contentId = contentSession.getContentId();

		final boolean recordOnRepository = contentId != null
				&& contentId.equalsIgnoreCase("repositoryRecorder");
		recorderUrl = contentSession.getHttpServletRequest().getScheme()
				+ "://" + config.getHandlerAddress() + ":"
				+ contentSession.getHttpServletRequest().getServerPort();
		handlerUrl = recorderUrl
				+ contentSession.getHttpServletRequest().getContextPath();
		cpbWindows = new CpbWindows(handlerUrl);

		mediaPipeline = contentSession.getMediaPipelineFactory().create();
		contentSession.releaseOnTerminate(mediaPipeline);

		rateLimiter = mediaPipeline.newGStreamerFilter(
				"videorate max-rate=15 average-period=200000000").build();
		mirrorFilter = mediaPipeline.newGStreamerFilter("videoflip method=4")
				.build();

		chromaFilter = mediaPipeline.newChromaFilter(
				new WindowParam(100, 10, 500, 400)).build();
		pointerDetectorAdvFilter = mediaPipeline.newPointerDetectorAdvFilter(
				new WindowParam(5, 5, 50, 50)).build();
		pointerDetectorAdvFilter.addWindow(cpbWindows.start);
		activeWindow = Windows.START;
		faceOverlayFilter = mediaPipeline.newFaceOverlayFilter().build();
		rateLimiter.connect(mirrorFilter);
		mirrorFilter.connect(pointerDetectorAdvFilter);
		pointerDetectorAdvFilter.connect(chromaFilter);
		chromaFilter.connect(faceOverlayFilter);

		pointerDetectorAdvFilter.addWindow(cpbWindows.fiware);
		pointerDetectorAdvFilter
				.addWindowInListener(new MediaEventListener<WindowInEvent>() {
					@Override
					public void onEvent(WindowInEvent event) {
						try {
							Windows windowId = Windows.valueOf(event
									.getWindowId());
							switch (windowId) {
							case DK:
								setDK();
								break;
							case FIWARE:
								break;
							case MARIO:
								setMario();
								break;
							case SF:
								setSF();
								break;
							case SONIC:
								setSonic();
								break;
							case START:
								addRecorder(contentSession, recordOnRepository);
								setStart();
								break;
							case TRASH:
							case YOUTUBE:
								setEnding(windowId, recordOnRepository);
								break;
							default:
								break;
							}
						} catch (URISyntaxException e) {
							e.printStackTrace();
						}
					}
				});
		WebRtcEndpoint webRtcEndpoint = mediaPipeline.newWebRtcEndpoint()
				.build();
		webRtcEndpoint.connect(rateLimiter);
		faceOverlayFilter.connect(webRtcEndpoint);
		contentSession.start(webRtcEndpoint);
	}

	@Override
	public ContentCommandResult onContentCommand(
			WebRtcContentSession contentSession, ContentCommand contentCommand)
			throws Exception {
		if (contentCommand.getType().equalsIgnoreCase("calibrate")) {
			pointerDetectorAdvFilter.trackColorFromCalibrationRegion();
		}
		return new ContentCommandResult(contentCommand.getData());
	}

	private void addRecorder(WebRtcContentSession contentSession,
			boolean repositoryRecorder) {
		itemId = "campus-party-" + Calendar.getInstance().getTimeInMillis();

		if (repositoryRecorder) {
			RepositoryItem repositoryItem = contentSession.getRepository()
					.createRepositoryItem(itemId);
			RepositoryHttpRecorder recorder = repositoryItem
					.createRepositoryHttpRecorder();
			getLogger().info(
					"repositoryRecorderUrl + recorder.getURL()" + recorderUrl
							+ recorder.getURL());
			recorderEndpoint = mediaPipeline.newRecorderEndpoint(
					recorderUrl + recorder.getURL()).build();
		} else {
			recorderEndpoint = mediaPipeline.newRecorderEndpoint(
					"file:///tmp/" + itemId).build();
		}
		faceOverlayFilter.connect(recorderEndpoint);
	}

	private void setStarWars() {
		faceOverlayFilter.setOverlayedImage(handlerUrl
				+ "/img/masks/darthvader.png", -0.5F, -0.5F, 1.7F, 1.7F);
		chromaFilter
				.setBackground(handlerUrl + "/img/background/deathstar.jpg");
	}

	private void setPirates() {
		faceOverlayFilter.setOverlayedImage(handlerUrl + "/img/masks/jack.png",
				-0.4F, -0.4F, 1.7F, 1.7F);
		chromaFilter.setBackground(handlerUrl + "/img/background/pirates.jpg");
	}

	private void setStart() {
		pointerDetectorAdvFilter.clearWindows();
		pointerDetectorAdvFilter.addWindow(cpbWindows.fiware);
		pointerDetectorAdvFilter.addWindow(cpbWindows.mario);
		pointerDetectorAdvFilter.addWindow(cpbWindows.dk);
		pointerDetectorAdvFilter.addWindow(cpbWindows.sf);
		pointerDetectorAdvFilter.addWindow(cpbWindows.sonic);
		recorderEndpoint.record();
	}

	private void checkFirstTime() {
		if (activeWindow.equals(Windows.START)) {
			pointerDetectorAdvFilter.addWindow(cpbWindows.trash);
			pointerDetectorAdvFilter.addWindow(cpbWindows.youtube);
		}
	}

	private boolean checkEasterEggs() {
		boolean isEasterEgg = false;
		if (count % 20 == 0) {
			// Each 20 times (20, 40, ...) Darth Vader hat/background is
			// shown
			setStarWars();
			isEasterEgg = true;
		} else if (count % 10 == 0) {
			// Each 10 times (10, 30, ...) the Jack Sparrow hat/background is
			// shown
			setPirates();
			isEasterEgg = true;
		}
		count++;
		return isEasterEgg;
	}

	private void setSF() throws URISyntaxException {
		if (!activeWindow.equals(Windows.SF) && !checkEasterEggs()) {
			faceOverlayFilter.setOverlayedImage(handlerUrl
					+ "/img/masks/sf.png", -0.35F, -0.5F, 1.6F, 1.6F);
			chromaFilter.setBackground(handlerUrl + "/img/background/sf.jpg");
			checkFirstTime();
			activeWindow = Windows.SF;
		}
	}

	private void setMario() throws URISyntaxException {
		if (!activeWindow.equals(Windows.MARIO) && !checkEasterEggs()) {
			chromaFilter
					.setBackground(handlerUrl + "/img/background/mario.jpg");
			// Mario Easter Egg (a different mask each time)
			if (mario % 2 == 0) {
				faceOverlayFilter.setOverlayedImage(handlerUrl
						+ "/img/masks/mario-wings.png", -0.35F, -1.2F, 1.6F,
						1.6F);
			} else {
				faceOverlayFilter.setOverlayedImage(handlerUrl
						+ "/img/masks/mario.png", -0.3F, -0.6F, 1.6F, 1.6F);
			}
			mario++;
			checkFirstTime();
			activeWindow = Windows.MARIO;
		}
	}

	private void setDK() throws URISyntaxException {
		if (!activeWindow.equals(Windows.DK) && !checkEasterEggs()) {
			faceOverlayFilter.setOverlayedImage(handlerUrl
					+ "/img/masks/dk.png", -0.35F, -0.5F, 1.6F, 1.6F);
			chromaFilter.setBackground(handlerUrl + "/img/background/dk.jpg");
			checkFirstTime();
			activeWindow = Windows.DK;
		}
	}

	private void setSonic() throws URISyntaxException {
		if (!activeWindow.equals(Windows.SONIC) && !checkEasterEggs()) {
			faceOverlayFilter.setOverlayedImage(handlerUrl
					+ "/img/masks/sonic.png", -0.5F, -0.5F, 1.7F, 1.7F);
			chromaFilter
					.setBackground(handlerUrl + "/img/background/sonic.jpg");
			checkFirstTime();
			activeWindow = Windows.SONIC;
		}
	}

	private void setEnding(Windows windowId, boolean recordOnRepository) {
		chromaFilter.unsetBackground();
		pointerDetectorAdvFilter.clearWindows();
		pointerDetectorAdvFilter.addWindow(cpbWindows.fiware);
		faceOverlayFilter.unsetOverlayedImage();
		pointerDetectorAdvFilter.addWindow(cpbWindows.start);
		activeWindow = Windows.START;
		recorderEndpoint.release();
		if (windowId.equals(Windows.YOUTUBE)) {
			String recordUrl = handlerUrl
					+ (recordOnRepository ? "/playerRepository/"
							: "/cpbPlayer/") + itemId;
			getLogger().info("recordUrl " + recordUrl);
			Videos.upload(
					recordUrl,
					PLAYLIST_TOKEN,
					newArrayList("FI-WARE", "Kurento", "FUN-LAB", "GSyC",
							"URJC", "Campus Party", "WebRTC",
							"Software Engineering", "Augmented Reality",
							"Computer Vision", "Super Mario", "Sonic",
							"Street Fighter", "Donkey Kong"));
		}
	}

}

class CpbWindows {

	// Enumeration of windows
	enum Windows {
		START, SF, DK, MARIO, SONIC, TRASH, YOUTUBE, FIWARE
	}

	// Windows instances
	public PointerDetectorWindowMediaParam start;
	public PointerDetectorWindowMediaParam sf;
	public PointerDetectorWindowMediaParam mario;
	public PointerDetectorWindowMediaParam dk;
	public PointerDetectorWindowMediaParam sonic;
	public PointerDetectorWindowMediaParam trash;
	public PointerDetectorWindowMediaParam youtube;
	public PointerDetectorWindowMediaParam fiware;

	public CpbWindows(String handlerUrl) throws URISyntaxException {
		start = new PointerDetectorWindowMediaParam(Windows.START.toString(),
				100, 100, 280, 380);
		start.setImage(handlerUrl + "/img/buttons/start.png");

		mario = new PointerDetectorWindowMediaParam(Windows.MARIO.toString(),
				100, 100, 540, 0);
		mario.setImage(handlerUrl + "/img/buttons/mario.png");

		dk = new PointerDetectorWindowMediaParam(Windows.DK.toString(), 100,
				100, 540, 126);
		dk.setImage(handlerUrl + "/img/buttons/dk.png");

		sf = new PointerDetectorWindowMediaParam(Windows.SF.toString(), 100,
				100, 540, 252);
		sf.setImage(handlerUrl + "/img/buttons/sf.png");

		sonic = new PointerDetectorWindowMediaParam(Windows.SONIC.toString(),
				100, 100, 540, 380);
		sonic.setImage(handlerUrl + "/img/buttons/sonic.png");

		youtube = new PointerDetectorWindowMediaParam(
				Windows.YOUTUBE.toString(), 100, 100, 0, 380);
		youtube.setImage(handlerUrl + "/img/buttons/youtube.png");

		trash = new PointerDetectorWindowMediaParam(Windows.TRASH.toString(),
				100, 100, 0, 190);
		trash.setImage(handlerUrl + "/img/buttons/trash.png");

		fiware = new PointerDetectorWindowMediaParam(Windows.FIWARE.toString(),
				40, 180, 230, 0);
		fiware.setImage(handlerUrl + "/img/buttons/fiware2.png");
	}
}

File: example/brazil_demo_2/cpbWebRtc.js
/*
 * (C) Copyright 2014 Kurento (http://kurento.org/)
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Lesser General Public License
 * (LGPL) version 2.1 which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/lgpl-2.1.html
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 */

var handlerUrl = 'https://raw.github.com/Kurento/kmf-content-demo/develop/src/main/webapp';

var Windows =
{
  DK:
  {
    id:            'DK',
    height:        100,
    width:         100,
    upperRightX:   540,
    upperRightY:   126,
    image:         handlerUrl+'/img/buttons/dk.png',
    inactiveImage: handlerUrl+'/img/buttons/dk.png'
  },

  FIWARE:
  {
    id:            'FIWARE',
    height:        40,
    width:         180,
    upperRightX:   230,
    upperRightY:   0,
    image:         handlerUrl+'/img/buttons/fiware2.png',
    inactiveImage: handlerUrl+'/img/buttons/fiware2.png'
  },

  MARIO:
  {
    id:            'MARIO',
    height:        100,
    width:         100,
    upperRightX:   540,
    upperRightY:   0,
    image:         handlerUrl+'/img/buttons/mario.png',
    inactiveImage: handlerUrl+'/img/buttons/mario.png'
  },

  SF:
  {
    id:            'SF',
    height:        100,
    width:         100,
    upperRightX:   540,
    upperRightY:   252,
    image:         handlerUrl+'/img/buttons/sf.png',
    inactiveImage: handlerUrl+'/img/buttons/sf.png'
  },

  SONIC:
  {
    id:            'SONIC',
    height:        100,
    width:         100,
    upperRightX:   540,
    upperRightY:   380,
    image:         handlerUrl+'/img/buttons/sonic.png',
    inactiveImage: handlerUrl+'/img/buttons/sonic.png'
  },

  START:
  {
    id:            'START',
    height:        100,
    width:         100,
    upperRightX:   280,
    upperRightY:   380,
    image:         handlerUrl+'/img/buttons/start.png',
    inactiveImage: handlerUrl+'/img/buttons/start.png'
  },

  TRASH:
  {
    id:            'TRASH',
    height:        100,
    width:         100,
    upperRightX:   0,
    upperRightY:   100,
    image:         handlerUrl+'/img/buttons/trash.png',
    inactiveImage: handlerUrl+'/img/buttons/trash.png'
  },

  YOUTUBE:
  {
    id:            'YOUTUBE',
    height:        100,
    width:         100,
    upperRightX:   380,
    upperRightY:   0,
    image:         handlerUrl+'/img/buttons/youtube.png',
    inactiveImage: handlerUrl+'/img/buttons/youtube.png'
  }
};

var PLAYLIST_TOKEN = "PL58tWS2XjtialwG-eWDYoFwQpHTd5vDEE";


function CpbWebRtc()
{
  var activeWindow = Windows.START;

  var count = 0;
  var mario = 0;

  var pointerDetectorAdvFilter = null;
  var chromaFilter             = null;
  var faceOverlayFilter        = null;
  var recorderEndpoint         = null;


  function checkEasterEgg()
  {
    count++;

    if(count % 20 == 0)
    {
      // Each 20 times (20, 40, ...) Darth Vader hat/background is shown
      setStarWars();

      return true;
    };

    if(count % 10 == 0)
    {
      // Each 10 times (10, 30, ...) the Jack Sparrow hat/background is shown
      setPirates();

      return true;
    };

    return false;
  };

  function checkFirstTime()
  {
    if(activeWindow === Windows.START)
    {
      pointerDetectorAdvFilter.addWindow(cpbWindows.trash);
      pointerDetectorAdvFilter.addWindow(cpbWindows.youtube);
    };
  };


  function setDK()
  {
    if(activeWindow === Windows.DK || checkEasterEggs()) return;

    chromaFilter.setBackground(handlerUrl + "/img/background/dk.jpg");
    faceOverlayFilter.setOverlayedImage(handlerUrl + "/img/masks/dk.png",
                                        -0.35, -0.5, 1.6, 1.6);

    checkFirstTime();
    activeWindow = Windows.DK;
  };

  function setMario()
  {
    if(activeWindow === Windows.MARIO || checkEasterEggs()) return;

    chromaFilter.setBackground(handlerUrl + "/img/background/mario.jpg");

    mario++;

    // Mario Easter Egg (a different mask each time)
    if(mario % 2 == 0)
      faceOverlayFilter.setOverlayedImage(handlerUrl + "/img/masks/mario-wings.png",
                                          -0.35, -1.2, 1.6, 1.6);
    else
      faceOverlayFilter.setOverlayedImage(handlerUrl + "/img/masks/mario.png",
                                          -0.3, -0.6, 1.6, 1.6);

    checkFirstTime();
    activeWindow = Windows.MARIO;
  };

  function setPirates()
  {
    chromaFilter.setBackground(handlerUrl + "/img/background/pirates.jpg");
    faceOverlayFilter.setOverlayedImage(handlerUrl + "/img/masks/jack.png",
                                        -0.4, -0.4, 1.7, 1.7);
  };

  function setSF()
  {
    if(activeWindow === Windows.SF || checkEasterEggs()) return;

    chromaFilter.setBackground(handlerUrl + "/img/background/sf.jpg");
    faceOverlayFilter.setOverlayedImage(handlerUrl + "/img/masks/sf.png",
                                        -0.35, -0.5, 1.6, 1.6);

    checkFirstTime();
    activeWindow = Windows.SF;
  };

  function setSonic()
  {
    if(activeWindow !== Windows.SONIC && !checkEasterEggs())
    {
      chromaFilter.setBackground(handlerUrl + "/img/background/sonic.jpg");
      faceOverlayFilter.setOverlayedImage(handlerUrl + "/img/masks/sonic.png",
                                          -0.5, -0.5, 1.7, 1.7);

      checkFirstTime();
      activeWindow = Windows.SONIC;
    };
  };

  function setStarWars()
  {
    chromaFilter.setBackground(handlerUrl + "/img/background/deathstar.jpg");
    faceOverlayFilter.setOverlayedImage(handlerUrl + "/img/masks/darthvader.png",
                                        -0.5, -0.5, 1.7, 1.7);
  };


  function setStart()
  {
    RecorderEndpoint.create(pipeline, function(error, endpoint)
    {
      if(error) return console.error(error);

      recorderEndpoint = endpoint;

      faceOverlayFilter.connect(recorderEndpoint);

      pointerDetectorAdvFilter.clearWindows();

      pointerDetectorAdvFilter.addWindow(Windows.DK);
      pointerDetectorAdvFilter.addWindow(Windows.FIWARE);
      pointerDetectorAdvFilter.addWindow(Windows.MARIO);
      pointerDetectorAdvFilter.addWindow(Windows.SF);
      pointerDetectorAdvFilter.addWindow(Windows.SONIC);

      recorderEndpoint.record();
    });
  };

  function setEnding(windowID)
  {
    chromaFilter.unsetBackground();

    pointerDetectorAdvFilter.clearWindows();
    pointerDetectorAdvFilter.addWindow(Windows.FIWARE);
    pointerDetectorAdvFilter.addWindow(Windows.START);

    faceOverlayFilter.unsetOverlayedImage();

    activeWindow = Windows.START;

    recorderEndpoint.release();

    if(windowID == Windows.YOUTUBE.id)
    {
      console.info('recordUrl ');


    };
  };


  function onWindowIn(windowID)
  {
    switch(windowID)
    {
      case 'DK':     setDK();     break;

      case 'FIWARE':              break;
      case 'MARIO':  setMario();  break;

      case 'SF':     setSF();     break;
      case 'SONIC':  setSonic();  break;
      case 'START':  setStart();  break;

      case 'TRASH':
      case 'YOUTUBE':  setEnding(windowID);  break;

      default:
        console.warning('Unknown windowID',windowID);
    };
  };


  // Create a new connection
  var contentId = document.getElementById("contentId");
      contentId = contentId.disabled ? "" : "/" + contentId.value;

  var endpoint = 'ws://kms01.kurento.org:8080/thrift/ws/websocket';
//  var endpoint = 'ws://kms01.kurento.org:8080/thrift/ws/websocket' + contentId;

  var kwsMedia = null;


  /**
   * Javascript port of the Campus Party Brazil 2014 Kurento demo
   *
   * This demo has the following pipeline:
   *
   * WebRTC -> RateLimiter -> MirrorFilter -> PointerDetectorFilter ->
   * ChromaFilter -> FaceOverlayFilter -> Recorder
   *
   * @author Jesús Leganés Combarro "piranna" (piranna@gmail.com)
   * @since 1.0.0
   */
  this.start = function(offer, callback)
  {
    kwsMedia = KwsMedia(endpoint, function(kwsMedia)
    {
      // Create pipeline
      kwsMedia.create('MediaPipeline', function(error, pipeline)
      {
        if(error) return callback(error);

        pipeline.create('WebRtcEndpoint', function(error, webRtcEndpoint)
        {
          if(error) return callback(error);

          // Create pipeline media elements (endpoints & filters)
          pipeline.create('RateFilter', function(error, rateFilter)
          {
            if(error) return callback(error);

            webRtcEndpoint.connect(rateFilter, function(error)
            {
              if(error) return callback(error);

              pipeline.create('MirrorFilter', function(error, mirrorFilter)
              {
                if(error) return callback(error);

                rateFilter.connect(mirrorFilter, function(error)
                {
                  if(error) return callback(error);

                  var calibrationRegion =
                  {
                    topRightCornerX: 5,
                    topRightCornerY: 5,
                    width: 50,
                    height: 50
                  };
                  var params = {calibrationRegion: calibrationRegion};

                  pipeline.create('PointerDetectorAdvFilter', params,
                  function(error, filter)
                  {
                    if(error) return callback(error);

                    pointerDetectorAdvFilter = filter;

                    pointerDetectorAdvFilter.addWindow(Windows.START,
                    function(error)
                    {
                      if(error) return callback(error);

                      pointerDetectorAdvFilter.addWindow(Windows.FIWARE,
                      function(error)
                      {
                        if(error) return callback(error);

                        pointerDetectorAdvFilter.on('WindowIn', onWindowIn);

                        mirrorFilter.connect(pointerDetectorAdvFilter,
                        function(error)
                        {
                          if(error) return callback(error);

                          var window =
                          {
                            topRightCornerX: 100,
                            topRightCornerY: 10,
                            width: 500,
                            height: 400
                          };

                          pipeline.create('ChromaFilter', {window: window},
                          function(error, filter)
                          {
                            if(error) return callback(error);

                            chromaFilter = filter;

                            pointerDetectorAdvFilter.connect(chromaFilter,
                            function(error)
                            {
                              if(error) return callback(error);

                              pipeline.create('FaceOverlayFilter',
                              function(error, filter)
                              {
                                if(error) return callback(error);

                                faceOverlayFilter = filter;

                                chromaFilter.connect(faceOverlayFilter,
                                function(error)
                                {
                                  if(error) return callback(error);

                                  // Loopback
                                  faceOverlayFilter.connect(webRtcEndpoint,
                                  function(error)
                                  {
                                    if(error) return callback(error);

                                    webRtcEndpoint.processOffer(offer.sdp, callback);
                                  });
                                });
                              });
                            });
                          });
                        });
                      });
                    });
                  });
                });
              });
            });
          });
        });
      });

      kwsMedia.on("error", callback);
    },
    callback);
  };

  this.terminate = function()
  {
    if(kwsMedia)
    {
      kwsMedia.close();
      kwsMedia = null;
    }
    else
      console.warn('kwsMedia is not initialized');
  };

  this.calibrate = function()
  {
    if(pointerDetectorAdvFilter)
       pointerDetectorAdvFilter.trackcolourFromCalibrationRegion(function(error)
       {
         if(error) console.error(error);
       });
    else
      console.warn('pointerDetectorAdvFilter is not initialized');
  };
};
